File: dp/cpp/half_sync_half_async.cpp
// Half-Sync/Half-Async — Concurrency Pattern (POSA2)
//
// Intent: separate the system into an asynchronous I/O layer (which uses
//         non-blocking calls or event notification) and a synchronous
//         processing layer (which uses blocking, easier-to-write code).
//         A thread-safe queue connects the two layers.
//
// Architecture:
//
//     [ async I/O thread ]   <- low level, non-blocking
//             |
//        [ shared queue ]    <- decouples the two halves
//             |
//     [ sync worker pool ]   <- high level, blocking, easy to write
//
// When to use:
//   - I/O is best done asynchronously (epoll, IOCP, io_uring)
//   - Business logic is easier to write synchronously
//   - You want to bound how many requests are processed concurrently
//
// When NOT to use:
//   - Both layers can be done with the same model (then just pick one)
//   - Latency-sensitive — the queue hand-off adds wakeup jitter

#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <queue>
#include <vector>
#include <chrono>
#include <atomic>
#include <string>

// ============================================================================
// Shared queue: the seam between the two halves
// ============================================================================
template <typename T>
class BlockingQueue {
public:
    void push(T item) {
        {
            std::lock_guard<std::mutex> lock(mtx_);
            q_.push(std::move(item));
        }
        cv_.notify_one();
    }
    bool pop(T& out) {
        std::unique_lock<std::mutex> lock(mtx_);
        cv_.wait(lock, [this] { return !q_.empty() || closed_; });
        if (q_.empty()) return false;
        out = std::move(q_.front());
        q_.pop();
        return true;
    }
    void close() {
        {
            std::lock_guard<std::mutex> lock(mtx_);
            closed_ = true;
        }
        cv_.notify_all();
    }

private:
    std::queue<T>           q_;
    std::mutex              mtx_;
    std::condition_variable cv_;
    bool                    closed_ = false;
};

// ============================================================================
// Async half: simulates an event-driven I/O layer
// ============================================================================
class AsyncReceiver {
public:
    AsyncReceiver(BlockingQueue<std::string>& queue) : queue_(queue) {}

    // In a real system this would be an epoll loop. Here we just generate
    // events on a timer.
    void run(int num_events) {
        for (int i = 0; i < num_events; ++i) {
            std::this_thread::sleep_for(std::chrono::milliseconds(8));
            std::string event = "request-" + std::to_string(i);
            std::cout << "[async] received " << event << "\n";
            queue_.push(std::move(event));
        }
        queue_.close();
    }

private:
    BlockingQueue<std::string>& queue_;
};

// ============================================================================
// Sync half: a small pool of workers doing blocking processing
// ============================================================================
class SyncWorkerPool {
public:
    SyncWorkerPool(BlockingQueue<std::string>& queue, std::size_t num_workers) {
        for (std::size_t i = 0; i < num_workers; ++i) {
            workers_.emplace_back([&queue, i] {
                std::string event;
                while (queue.pop(event)) {
                    // Real work — slow, blocking, but easy to write synchronously
                    std::this_thread::sleep_for(std::chrono::milliseconds(20));
                    std::cout << "  [worker " << i << "] processed " << event << "\n";
                }
            });
        }
    }

    ~SyncWorkerPool() {
        for (auto& t : workers_) {
            if (t.joinable()) t.join();
        }
    }

private:
    std::vector<std::thread> workers_;
};

int main() {
    std::cout << "=== Half-Sync/Half-Async Pattern ===\n";
    BlockingQueue<std::string> queue;

    // Sync workers start first, blocking on the queue
    SyncWorkerPool workers(queue, 3);

    // Async receiver runs and feeds the queue
    AsyncReceiver receiver(queue);
    receiver.run(8);

    // workers join via destructor

    std::cout << "\nKey points:\n";
    std::cout << " * Async layer handles non-blocking I/O — high throughput\n";
    std::cout << " * Sync layer can do blocking work — simple to write\n";
    std::cout << " * Queue is the only shared state — clean isolation\n";
    std::cout << " * Each layer can be tuned (thread counts) independently\n";
    return 0;
}

File: dp/rust/src/half_sync_half_async.rs
// Half-Sync/Half-Async Pattern in Rust
// Async layer queues work, sync layer processes it

use std::sync::mpsc;
use std::thread;
use std::time::Duration;

#[derive(Debug)]
struct Request {
    id: u32,
    payload: String,
}

#[derive(Debug)]
struct Response {
    request_id: u32,
    result: String,
}

// Async layer: accepts requests without blocking
struct AsyncLayer {
    sender: mpsc::Sender<Request>,
}

impl AsyncLayer {
    fn submit(&self, req: Request) {
        println!("[Async] Queuing request {}", req.id);
        self.sender.send(req).unwrap();
    }
}

// Sync layer: processes requests synchronously in worker threads
fn sync_worker(id: usize, rx: mpsc::Receiver<Request>, result_tx: mpsc::Sender<Response>) {
    while let Ok(req) = rx.recv() {
        println!("[Sync Worker {}] Processing request {}", id, req.id);
        // Simulate blocking work
        thread::sleep(Duration::from_millis(100));
        let resp = Response {
            request_id: req.id,
            result: format!("Processed '{}' by worker {}", req.payload, id),
        };
        result_tx.send(resp).unwrap();
    }
}

fn main() {
    let (req_tx, req_rx) = mpsc::channel::<Request>();
    let (res_tx, res_rx) = mpsc::channel::<Response>();

    let async_layer = AsyncLayer { sender: req_tx };

    // Start sync workers
    let req_rx = std::sync::Arc::new(std::sync::Mutex::new(req_rx));
    let workers: Vec<_> = (0..2)
        .map(|id| {
            let rx = std::sync::Arc::clone(&req_rx);
            let tx = res_tx.clone();
            thread::spawn(move || {
                while let Ok(req) = rx.lock().unwrap().recv() {
                    println!("[Sync Worker {}] Processing request {}", id, req.id);
                    thread::sleep(Duration::from_millis(80));
                    tx.send(Response {
                        request_id: req.id,
                        result: format!("Done '{}' by worker {}", req.payload, id),
                    })
                    .unwrap();
                }
            })
        })
        .collect();
    drop(res_tx);

    // Async layer submits requests
    for i in 0..5 {
        async_layer.submit(Request {
            id: i,
            payload: format!("task-{}", i),
        });
    }
    drop(async_layer);

    // Collect results
    while let Ok(resp) = res_rx.recv() {
        println!("[Result] Request {}: {}", resp.request_id, resp.result);
    }

    for w in workers {
        w.join().unwrap();
    }
    println!("All done.");
}
